### ISSUE TITLE
tesstrain continues on text2image error

### ISSUE THREAD

**Issue by Shreeshrii** [tesseract-ocr/tesseract#2313]
```
Stripped 5 unrenderable words
Error in pixCreateNoInit: pix_malloc fail for data
Error in pixCreateTemplateNoInit: pixd not made
Error in pixGetData: pix not defined
Error in pixGetWpl: pix not defined
Error in composeRGBPixel: &pixel not defined
Rendered page 56 to file /tmp/tam-2019-03-12.PQP/tam.Noto_Sans_Tamil_Bold.exp0.tif
Stripped 1 unrenderable words
ERROR: Program text2image failed. Abort.
Rendered page 57 to file /tmp/tam-2019-03-12.PQP/tam.TAMu_Kadambri_Regular.exp0.tif
Stripped 50 unrenderable words
Rendered page 57 to file /tmp/tam-2019-03-12.PQP/tam.Noto_Sans_Tamil_Bold.exp0.tif
Stripped 3 unrenderable words
ERROR: Program text2image failed. Abort.
```

Even on error the bash script continues and dies only at end of phase:

```
Stripped 2 unrenderable words
Rendered page 117 to file /tmp/tam-2019-03-12.PQP/tam.e-Grantamil.exp0.tif
Rendered page 118 to file /tmp/tam-2019-03-12.PQP/tam.e-Grantamil.exp0.tif
Rendered page 119 to file /tmp/tam-2019-03-12.PQP/tam.e-Grantamil.exp0.tif
ERROR: /tmp/tam-2019-03-12.PQP/tam.Arial_Unicode_MS.exp0.box does not exist or is not readable
```

*Expected* - End script at the first error itself so that problem can be investigated early.

**Comment by zdenop** [comment]
Please provide instruction/test case how to generate this problem...

**Comment by Shreeshrii** [comment]
I updated to latest code from master branch without any additional PRs. Now  I get the following error:

```
Rendered page 27 to file /tmp/tam-2019-03-12.p1P/tam.Arial_Unicode_MS.exp0.tif
Rendered page 26 to file /tmp/tam-2019-03-12.p1P/tam.Lohit_Tamil_Regular.exp0.tif
Stripped 9 unrenderable words
Rendered page 27 to file /tmp/tam-2019-03-12.p1P/tam.FreeSerif.exp0.tif
Stripped 4 unrenderable words
/home/ubuntu/tesseract/src/training/tesstrain_utils.sh: line 71: 19898 Killed                  "${cmd}" "$@" 2>&1 1>&2
     19899 Done                    | tee -a ${LOG_FILE}
Rendered page 27 to file /tmp/tam-2019-03-12.p1P/tam.Karla_Tamil_Inclined_Italic.exp0.tif
Stripped 23 unrenderable words
Rendered page 27 to file /tmp/tam-2019-03-12.p1P/tam.Karla_Tamil_Upright_Regular.exp0.tif
```

**Comment by stweil** [comment]
"Killed" might be the result of one of the new assertions which I added to detect use of a wrong unicharset. Run `ulimit -c unlimited` in the Linux shell before starting the training in the same shell. Then a crash caused by an assertion or other reason will create a core dump file named `core` or `core*`. `file core` typically shows which executable crashed. `gdb EXECUTABLE core` can be used to show debug information for the crash, for example stack trace or values of variables.

**Comment by Shreeshrii** [comment]
@stweil Where will the `core` be created? tesstrain.sh is continuing with these errors too.

>/home/ubuntu/tesseract/src/training/tesstrain_utils.sh: line 71: 19898 Killed
>                  "${cmd}" "$@" 2>&1 1>&2

What about the changes in the commit https://github.com/tesseract-ocr/tesseract/pull/2294/commits/91a7e86bf5e90fc267e921748f4412c553b9e87e

**Comment by stweil** [comment]
The core file is normally created in the directory which is the current directory at the time of the crash (important when the crashing executable was called from a script).

I think Robert's change is important for you, too. Maybe I should merge it separately before the rest of PR #2294 is ready to get merged. cc @bertsky.

**Comment by Shreeshrii** [comment]
@stweil No `core` file is generated. In the case of these errors, text2image does not crash or give any error message - rather the error is generated in the bash script by checking whether a box file has been created or not. Usually only a partial tif file is generated for these fonts.

This is happening in cases of processing text2image with --distort_image=true and using tesstrain.sh which by default runs 8 parallel processes for generating images. If I change it to process only one image at a time (`phase_I_generate_image 1`)  then the files are created ok without error, though the script takes much longer to run.

**Comment by bertsky** [comment]
Well I do think that 91a7e86 is also relevant here, but there is more to it: the parallelism in `tesstrain_utils.sh` for [`generate_font_image`](https://github.com/tesseract-ocr/tesseract/blob/4c2bbebeccf28bac747fa1e567893906ccdfe8a2/src/training/tesstrain_utils.sh#L319) does not check the return status of its background jobs (despite `errexit` being inherited by subshells in non-POSIX mode). That would be `wait -n` or so. From the bash manual:

> Subshells spawned to execute command substitutions inherit the value of the -e option from the parent shell.  When not in posix mode, bash clears the -e option in such subshells.

Should I open a new PR or can I add to #2294 ?

**Comment by Shreeshrii** [comment]
>Should I open a new PR or can I add to #2294 ?

New PR, please!

**Comment by bertsky** [comment]
Won't that complicate merging? After all, this does depend on other unmerged changes to the same file contained in #2294. Should I try to isolate both changesets?

**Comment by Shreeshrii** [comment]
@stweil Even though I couldn't find a core file, /var/logs/kern.log has the following messages:

```
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291459] text2image invoked oom-killer: gfp_mask=0x24200ca, order=0, oom_score_adj=0
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291465] text2image cpuset=/ mems_allowed=0
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291474] CPU: 1 PID: 4517 Comm: text2image Not tainted 4.4.0-142-generic #168-Ubuntu
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291476] Call Trace:
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291484] [c0000003f1ce77c0] [c000000000b23098] dump_stack+0xb0/0xf0 (unreliable)
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291488] [c0000003f1ce7800] [c000000000b20564] dump_header+0x84/0x210
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291493] [c0000003f1ce78d0] [c000000000233bc8] oom_kill_process+0x478/0x540
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291495] [c0000003f1ce7990] [c000000000233f90] out_of_memory+0x240/0x5f0
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291497] [c0000003f1ce7a40] [c00000000023cbe0] __alloc_pages_nodemask+0xf50/0xf90
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291501] [c0000003f1ce7c20] [c0000000002a6a58] alloc_pages_vma+0xe8/0x350
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291504] [c0000003f1ce7cb0] [c00000000027792c] handle_mm_fault+0x114c/0x1940
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291506] [c0000003f1ce7d80] [c000000000b19e84] do_page_fault+0x364/0x820
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291509] [c0000003f1ce7e30] [c0000000000087e4] handle_page_fault+0x10/0x30
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291510] Mem-Info:
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291517] active_anon:255255 inactive_anon:1973 isolated_anon:0
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291517]  active_file:21 inactive_file:2 isolated_file:0
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291517]  unevictable:69 dirty:0 writeback:0 unstable:0
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291517]  slab_reclaimable:749 slab_unreclaimable:2070
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291517]  mapped:65 shmem:2323 pagetables:306 bounce:0
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291517]  free:254 free_pcp:26 free_cma:0
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291521] Node 0 DMA free:16256kB min:16256kB low:20288kB high:24384kB active_anon:16336320kB inactive_anon:126272kB active_file:1344kB inactive_file:128kB unevictable:4416kB isolated(anon):0kB isolated(file):0kB present:16777216kB managed:16738944kB mlocked:4416kB dirty:0kB writeback:0kB mapped:4160kB shmem:148672kB slab_reclaimable:47936kB slab_unreclaimable:132480kB kernel_stack:3184kB pagetables:19584kB unstable:0kB bounce:0kB free_pcp:1664kB local_pcp:192kB free_cma:0kB writeback_tmp:0kB pages_scanned:62144 all_unreclaimable? yes
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291528] lowmem_reserve[]: 0 0 0 0
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291530] Node 0 DMA: 68*64kB (UME) 27*128kB (UME) 24*256kB (UME) 0*512kB 0*1024kB 0*2048kB 0*4096kB 0*8192kB 0*16384kB = 13952kB
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291538] 2414 total pagecache pages
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291540] 0 pages in swap cache
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291542] Swap cache stats: add 0, delete 0, find 0/0
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291543] Free swap  = 0kB
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291544] Total swap = 0kB
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291545] 262144 pages RAM
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291546] 0 pages HighMem/MovableOnly
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291547] 598 pages reserved
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291548] 0 pages cma reserved
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291549] 0 pages hwpoisoned
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291550] [ pid ]   uid  tgid total_vm      rss nr_ptes nr_pmds swapents oom_score_adj name
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291554] [  661]     0   661      370       84       6       3        0             0 systemd-journal
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291556] [  690]     0   690     1250       30       6       3        0             0 lvmetad
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291559] [  742]     0   742      256       41       6       3        0         -1000 systemd-udevd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291561] [ 2268]   100  2268     1242       31       6       3        0             0 systemd-timesyn
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291563] [ 2421]     0  2421      118       50       5       3        0             0 dhclient
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291565] [ 2540]     0  2540      120       39       5       3        0             0 dhclient
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291567] [ 2670]   104  2670     3578       57       8       3        0             0 rsyslogd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291569] [ 2672]     0  2672     3527       32       9       3        0             0 lxcfs
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291571] [ 2674]     0  2674     3712       69       9       3        0             0 accounts-daemon
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291573] [ 2683]     0  2683       89       44       4       3        0             0 atd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291575] [ 2685]     0  2685      136       31       5       3        0             0 cron
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291578] [ 2691]     0  2691     1325        8       5       2        0             0 iprdump
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291580] [ 2692]     0  2692      103       55       5       3        0             0 systemd-logind
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291582] [ 2700]     0  2700       46       20       4       3        0             0 acpid
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291584] [ 2702]   107  2702      149       45       4       3        0          -900 dbus-daemon
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291586] [ 2715]     0  2715     9392      243      23       5        0          -900 snapd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291588] [ 2727]     0  2727      236       84       5       3        0         -1000 sshd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291590] [ 2729]     0  2729       63       11       4       3        0             0 iscsid
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291592] [ 2730]     0  2730       71       63       4       3        0           -17 iscsid
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291594] [ 2749]     0  2749       48        9       4       3        0             0 iprupdate
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291596] [ 2750]     0  2750       48       10       4       3        0             0 iprinit
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291599] [ 2789]     0  2789      105       43       4       3        0             0 agetty
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291601] [ 2790]     0  2790      109       39       5       3        0             0 agetty
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291603] [ 2799]     0  2799       68       18       4       3        0             0 mdadm
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291605] [ 2801]     0  2801      106       45       5       3        0             0 irqbalance
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291607] [ 2802]     0  2802       84       46       5       3        0             0 rtas_errd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291609] [ 2810]     0  2810     3674       70       9       3        0             0 polkitd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291611] [  960]  1000   960      124       42       4       3        0             0 bash
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291613] [22602]     0 22602      284      100       5       3        0             0 sshd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291615] [22661]  1000 22661      284       93       5       3        0             0 sshd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291618] [22662]  1000 22662       60       41       4       3        0             0 sftp-server
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291620] [22665]     0 22665      284      100       5       3        0             0 sshd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291622] [22698]  1000 22698      284       93       5       3        0             0 sshd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291624] [22699]  1000 22699      163       78       5       3        0             0 bash
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291626] [22953]     0 22953      284      101       6       3        0             0 sshd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291628] [22986]  1000 22986      287      114       6       3        0             0 sshd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291630] [22987]  1000 22987       60       38       5       3        0             0 sftp-server
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291632] [24593]     0 24593      284      101       5       3        0             0 sshd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291634] [24626]  1000 24626      284      112       5       3        0             0 sshd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291636] [24627]  1000 24627      123       41       5       3        0             0 bash
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291638] [32351]  1000 32351      124       37       5       3        0             0 bash
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291640] [ 1733]     0  1733      284      123       5       3        0             0 sshd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291642] [ 1788]  1000  1788      284      105       5       3        0             0 sshd
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291644] [ 1789]  1000  1789      163       50       5       3        0             0 bash
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291647] [ 4264]  1000  4264     4406     2941      19       3        0             0 lt-lstmeval
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291649] [ 4504]  1000  4504      134       39       4       3        0             0 bash
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291651] [ 4517]  1000  4517   124367   122068     488       3        0             0 text2image
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291653] [ 4518]  1000  4518       93        7       4       3        0             0 tee
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291655] [ 4558]  1000  4558      134       39       4       3        0             0 bash
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291657] [ 4571]  1000  4571   124355   121991     489       3        0             0 text2image
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291659] [ 4572]  1000  4572       93        7       4       3        0             0 tee
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291661] [ 4580]  1000  4580     7773     5984      33       3        0             0 lt-lstmeval
Mar 14 15:04:16 tesseract-ocr kernel: [249079.291663] Out of memory: Kill process 4517 (text2image) score 468 or sacrifice child
Mar 14 15:04:16 tesseract-ocr kernel: [249079.292894] Killed process 4517 (text2image) total-vm:7959488kB, anon-rss:7810112kB, file-rss:2240kB
```

**Comment by stweil** [comment]
So we know now that `text2image` was killed because it used excessive memory. This kind of failure won't create a core dump. The training process runs several `text2image` processes in parallel (2 in the listing above). Maybe you don't have enough memory for that. Then either the processes should run sequentially (one at a time), or you need more memory / swap space.

**Comment by Shreeshrii** [comment]
@stweil That is right. I also guessed so when the script had worked with (phase_I_generate_image 1).

However, the issue is that if eg. the error is on the first font of 40 fonts given for training, tesstrain.sh stops only after all 40 have been processed. Even though we get the message 'Killed' the bash script continues.

I haven't tried the python script yet. Need to figure out how to use it.

**Comment by Shreeshrii** [comment]
@stweil I have now added a swapfile. Thanks for the suggestion.

ubuntu@tesseract-ocr:~$ free -h
              total        used        free      shared  buff/cache   available
Mem:            15G        147M         15G        110M        461M         15G
Swap:          1.0G          0B        1.0G
ubuntu@tesseract-ocr:~$ swapon --show
NAME      TYPE  SIZE USED PRIO
/swapfile file 1024M   0B   -1

**Comment by Shreeshrii** [comment]
```
Rendered page 35 to file /tmp/tam-2019-03-15.8Fp/tam.FreeSerif.exp0.tif
Rendered page 32 to file /tmp/tam-2019-03-15.8Fp/tam.Latha.exp0.tif
Stripped 2 unrenderable words
Rendered page 32 to file /tmp/tam-2019-03-15.8Fp/tam.Kavivanar.exp0.tif
Stripped 7 unrenderable words
/home/ubuntu/tesseract/src/training/tesstrain_utils.sh: line 62:  3636 Killed                  "${cmd}" "$@" 2>&1
      3637 Done                    | tee -a ${LOG_FILE}
++ basename /usr/local/bin/text2image
+ err_exit 'Program text2image failed. Abort.'
+ test -z /tmp/tam-2019-03-15.8Fp/tesstrain.log
+ echo -e 'ERROR: Program' text2image failed. Abort.
+ tee -a /tmp/tam-2019-03-15.8Fp/tesstrain.log
Rendered page 33 to file /tmp/tam-2019-03-15.8Fp/tam.Latha.exp0.tif
ERROR: Program text2image failed. Abort.
+ exit 1
ubuntu@tesseract-ocr:~/tesstutorial$ Rendered page 34 to file /tmp/tam-2019-03-15.8Fp/tam.Latha.exp0.tif
Stripped 2 unrenderable words
Rendered page 33 to file /tmp/tam-2019-03-15.8Fp/tam.Kavivanar.exp0.tif
Stripped 14 unrenderable words
Rendered page 34 to file /tmp/tam-2019-03-15.8Fp/tam.Kavivanar.exp0.tif
```

exit 1 - but does not end the script.

**Comment by Shreeshrii** [comment]
I upgraded python to 3.6 an ran tesstrain.py. As far as I can tell, it runs all fonts in parallel (not just 8) and also continue error. cc: @jbarlow83 

```
Rendered page 71 to file tmp/tam-2019-03-15x3hrczim/tam.ETTamilNew.exp0.tif
Stripped 19 unrenderable words
Rendered page 72 to file tmp/tam-2019-03-15x3hrczim/tam.ETTamilNew.exp0.tif
Stripped 10 unrenderable words
Rendered page 73 to file tmp/tam-2019-03-15x3hrczim/tam.ETTamilNew.exp0.tif
Stripped 16 unrenderable words

[13:12:43] CRITICAL - Program /usr/local/bin/text2image failed with return code -9. Abort.
[13:12:55] ERROR - Stripped 8 unrenderable words
Rendered page 0 to file tmp/tam-2019-03-15x3hrczim/tam.Baloo_Thambi.exp0.tif
Stripped 6 unrenderable words
Rendered page 1 to file tmp/tam-2019-03-15x3hrczim/tam.Baloo_Thambi.exp0.tif
Stripped 4 unrenderable words
```

**Comment by jbarlow83** [comment]
You can edit the script to set a limit on the number of workers.

At line 382:

```
        ) as pbar, concurrent.futures.ThreadPoolExecutor(max_workers=4) as executor:
```

I apologize but I won't be able to look into issues with the script till April - too many projects on the go.

**Comment by Shreeshrii** [comment]
@jbarlow83 Thanks for the hint. I have submitted a PR - https://github.com/tesseract-ocr/tesseract/pull/2335

**Comment by mohdbm** [comment]
is this issue resolved, because I get the same issue?!
